[JSC] GreedyRegAlloc: introduce gap points between instructions

dhecht opened the pull request:
#### bd03ee92315ed0b6eab52869d3665fd4f970581f
<pre>
[JSC] GreedyRegAlloc: introduce gap points between instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=302815">https://bugs.webkit.org/show_bug.cgi?id=302815</a>
<a href="https://rdar.apple.com/165070452">rdar://165070452</a>

Reviewed by Yusuke Suzuki.

Currently, there are two points per instruction: Early point which models
early effects and late point which models late effects. This means that
in the case of an early def by the first instruction in a block
or late use by the last instruction in a block, the live ranges
cannot distinguish whether the range is live into / out of
the block since the first / last points will include the ranges in either
case of the last point def&apos;ing / using the Tmp or the Tmp being live
across block boundaries.

Instead, let&apos;s use four points per instruction. Then, for the first
instruction of a block, the Pre point will be included in an interval
if and only if the Tmp is live into the block, and likewise for the
Post point and the block exit.

This will allow for more sophisticated live range splitting to be
implemented without needing to keep around explicit liveness information
for the duration of register allocation by allowing the live ranges to
be the full source of truth.

Bonus: Similarly, because the previous instruction&apos;s late point and next
instruction&apos;s early point are adjacent, the live range intervals
do not indicate kill positions in the case of a late use followed
by early def. Note however that this case is currently handled using
padInterference which inserts Nop instructions at these locations.
So as a bonus, this change removes a dependency on padInterference and
eventually padInterference can go away.

Canonical link: <a href="https://commits.webkit.org/304164@main">https://commits.webkit.org/304164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b96cf5de8225613adabc89120f12d6f8299329a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134681 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142205 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86619 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/59ba40fe-815f-4acd-aeb6-b21f5e47e542) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6989 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102938 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70214 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b37a062b-6e70-4fdc-b732-f8f6118a4f3c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5446 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120712 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83743 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/abed50ef-d549-45c9-8026-1f6b5fd9e144) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5298 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2900 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2795 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126726 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114525 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144899 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133189 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6810 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39434 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111332 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111627 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5120 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116986 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60694 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20800 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6861 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35177 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166098 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6665 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70442 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43414 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6901 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6774 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -128,6 +126,20 @@ static bool verbose() { return Options::airGreedyRegAllocVerbose(); }
 using Interval = WTF::Range<Point>;
 
+enum PointOffsets {
+    Pre = 0,
```
I am a bit confused. Pre is unused, but post is used by positionOfTail?

dhecht replied:
A block's positionOfHead will be a pre point and a block's positionOfTail will be a post point. Also, between each instruction is a post point (from the prev inst) and pre point (from the next inst). This bookkeeping change isn't too interesting by itself, but allows live ranges to be more precise about liveness on instruction boundaries (including block entry and exit) in the somewhat edge cases of early def and late uses. A future patch that does more live range splitting will depend on this.

dhecht committed:
[JSC] GreedyRegAlloc: introduce gap points between instructions
https://bugs.webkit.org/show_bug.cgi?id=302815
rdar://165070452

Reviewed by Yusuke Suzuki.

Currently, there are two points per instruction: Early point which models
early effects and late point which models late effects. This means that
in the case of an early def by the first instruction in a block
or late use by the last instruction in a block, the live ranges
cannot distinguish whether the range is live into / out of
the block since the first / last points will include the ranges in either
case of the last point def'ing / using the Tmp or the Tmp being live
across block boundaries.

Instead, let's use four points per instruction. Then, for the first
instruction of a block, the Pre point will be included in an interval
if and only if the Tmp is live into the block, and likewise for the
Post point and the block exit.

This will allow for more sophisticated live range splitting to be
implemented without needing to keep around explicit liveness information
for the duration of register allocation by allowing the live ranges to
be the full source of truth.

Bonus: Similarly, because the previous instruction's late point and next
instruction's early point are adjacent, the live range intervals
do not indicate kill positions in the case of a late use followed
by early def. Note however that this case is currently handled using
padInterference which inserts Nop instructions at these locations.
So as a bonus, this change removes a dependency on padInterference and
eventually padInterference can go away.

Canonical link: https://commits.webkit.org/304164@main
